packages/@aws-cdk/toolkit-lib/lib/actions/refactor/index.ts

```diff
@@ -142,3 +144,28 @@ export function parseMappingGroups(s: string) {
     }
   }
 }
+
+export interface EnvironmentSpecificMappings {
+  readonly environment: cxapi.Environment;
+  readonly mappings: Record<string, string>;
+}
+
+export async function mappingsByEnvironment(
+  stackArtifacts: cxapi.CloudFormationStackArtifact[],
+  sdkProvider: SdkProvider,
+  ignoreModifications?: boolean,
+): Promise<EnvironmentSpecificMappings[]> {
+  const groups = await groupStacks(sdkProvider, stackArtifacts, []);
+  return groups.map((group) => {
+    const context = new RefactoringContext({
+      ...group,
+      ignoreModifications,
+    });
+    return {
+      environment: context.environment,
+      mappings: Object.fromEntries(
```

Comment:
Rather than splatting down the object to a string here, aren't we better off maintaining the object and rendering it later when it comes time to present the diff?